methods_exploration/it_functions.py

```diff
@@ -99,3 +126,98 @@ def calcMI_crit(M, nbins = 11, alpha = 0.05, numiter = 1000, ncores = 2):
+    the mutual information is normalized by the entropy of the sink'''
+    #lag data
+    M_lagged = lag_data(M,shift)
+    #remove any rows where there is an nan value
+    M_short =  M_lagged[~np.isnan(M_lagged).any(axis=1)]
+    
+    M1 = M_short[:,(0,2)]  # [source_lagged(1:n-shift), sink_lagged(1:n-shift)]  =>H(Xt-T,Yt-T)
+    M2 = M_short[:,(1,2)] # [sink_unlagged(shift:n), sink_lagged(1:n-shift)]    =>H(Yt,Yt-T)
+    M3 = M_short[:,2]      # [sink_unlagged(1:n-shift)] =>H(Yt-T) 
+    
+    #calc joint entropy of H(Xt-T,Yt-T)
+    _, _, p_xlyl = calc2Dpdf(M1, nbins)
+    T1 = calcEntropy(p_xlyl)
+    
+    #calc joint entropy of H(Yt-t) and H(Yt,Yt-T)
+    _, pyl, p_yulyl = calc2Dpdf(M2, nbins)
+    T2 = calcEntropy(p_yulyl)
+    
+    #calc entropy of H(Yt-T)
+    T3 = calcEntropy(pyl)
+    
+    #calc 3d joint entropy 
+    p_xlyulyl = calc3Dpdfs(M_short, nbins)
+    T4 = calcEntropy(p_xlyulyl)
+
+    T = (T1+T2-T3-T4)/T3 # Knuth formulation of transfer entropy
+    
+    return T
+
+def calcTE_shuffled(M, shift, nbins = 11):
+    '''shuffles the input dataset to destroy temporal relationships for signficance testing
+    M: a numpy array of shape (nobs, 2) where nobs is the number of observations
+    this assumes that the data are arrange such that the first column is the source and
+    the second column is the sink.
+    nbins: is the number of bins used for estimating the pdf with a default of 11
+    the transfer entropy is normalized by the entropy of the sink
+    returns a single TE value for a numpy array the same size and shape as M, 
+    but with the order shuffled'''
+    
+    Mss = np.ones(np.shape(M))*np.nan # Initialize
+    
+    for n in range(np.shape(M)[1]): # Columns are shuffled separately
+        n_nans = np.argwhere(~np.isnan(M[:,n]))
+        R = np.random.rand(np.shape(n_nans)[0],1)
+        I = np.argsort(R,axis=0)
+        Mss[n_nans[:,0],n] = M[n_nans[I[:],0],n].reshape(np.shape(M[n_nans[I[:],0],n])[0],)
+    TE_shuff = calcTE(Mss, shift, nbins = 11)
+    return TE_shuff
+
+def calcTE_crit(M, shift, nbins = 11, alpha = 0.05, numiter = 1000, ncores = 2):
+    '''calculate the critical threshold of transfer entropy
+    M: a numpy array of shape (nobs, 2) where nobs is the number of observations
+    this assumes that the data are arrange such that the first column is the source and
+    the second column is the sink.
+    shift: time lag that should be considered
+    nbins: is the number of bins used for estimating the pdf with a default of 11
+    the transfer entropy is normalized by the entropy of the sink
+    alpha: sigficance threshold, default = 0.05
+    numiter: number of iterations, default = 1000
+    ncores: number of cores, default = 2'''
+    
+    TEss = Parallel(n_jobs=ncores)(delayed(calcTE_shuffled)(M, shift, nbins) for ii in range(numiter))
+    TEss = np.sort(TEss)
+    #print(MIss)
+    TEcrit = TEss[round((1-alpha)*numiter)] 
```

Comment:
I changed it to ceiling, not sure about returning the alpha value as that would just be `alpha`*`numiter`, so it will always be the 950th value when the shuffled transfer entropies are ordered (if `alpha = 0.05` and `numiter = 1000`). Maybe I am misunderstanding.

Reply:
I meant for a case like `numiter = 100` and `alpha = 0.005`. Then ceiling would return entry 100, which would correspond to a true alpha = 0, which I think would indicate needing to increase the `numiter` to estimate the critical value. Ceiling also helps if `numiter*(1-alpha)` doesn't return a whole number